draft-ietf-ppm-dap.md

```diff
@@ -1765,14 +1817,17 @@ Otherwise, the Leader proceeds as follows with each report:
 
 1. Else if the type is "finished" and `state == Finished(out_share)`, then
    preparation is complete and the Leader stores the output share for use in
-   the collection flow ({{collect-flow}}).
+   the collection interaction ({{collect-flow}}).
 
 1. Else if the type is "reject", then the Leader rejects the report and removes
    it from the candidate set.
 
 1. Else the type is invalid, in which case the Leader MUST abort the
    aggregation job.
 
+When the Leader stores the `out_share`, it MUST also store the report ID for
```

Comment:
I think we want to both check & store here, rather than only storing here. (this may have been the intent)

Otherwise, a Leader which checks at time of creation but only stores at time of completion could:
1) Generate two distinct aggregation jobs, both containing some report R. Both will check R against the report ID set, but will pass since R has not yet been stored in the set.
2) Concurrently step both aggregation jobs to completion. The Helper _should_ reject at least one of the copies of R as a replay; but if it doesn't, the Leader would store the report ID twice, rejecting neither copy of R.

That is, we are potentially left open to replay if the Helper's check is mis-implemented.

Reply:
The intent was only to store here. The only time the Leader checks is just after creating the aggregation job. It's therefore the Leader's responsibility to make sure it doesn't accidentally replay a report.

On the other hand, it's also the responsibility of the Helper to make sure a report isn't replayed. In other words, the scenario you're describing here is handled the same as an attack by a malicious Leader. I don't think we need to split hairs here.

Reply:
Yeah -- aggregation job creation & replay checking are similar, but have slightly different requirements.

A Leader doesn't want to put a report into more than one aggregation job, since that could lead to wasted work & more aggregation jobs for the same number of reports. But both the Leader's & Helper's replay checking should ensure that if a report does end up in two aggregation jobs, it will still only be aggregated once. So a Leader will want to track a couple of bits of state for each report: "is currently undergoing aggregation" (to control whether it is might be added to a new aggregation job) and "is aggregated into some batch" (to control whether it is considered a replay). These two bits are different since "is currently undergoing aggregation" might go back to false after being true (e.g. if it is `report_too_early`), but "is aggregated into some batch" is a one-way transition.